Sources/NIOCore/AsyncChannel/AsyncChannelInboundStreamChannelHandler.swift

```diff
@@ -135,20 +135,19 @@ internal final class NIOAsyncChannelInboundStreamChannelHandler<InboundIn: Senda
         case .transformation(let channelReadTransformation):
             // The unsafe transfers here are required because we need to use self in whenComplete
             // We are making sure to be on our event loop so we can safely use self in whenComplete
-            let unsafeSelf = NIOLoopBound(self, eventLoop: context.eventLoop)
-            let unsafeContext = NIOLoopBound(context, eventLoop: context.eventLoop)
             channelReadTransformation(unwrapped)
                 .hop(to: context.eventLoop)
+                .assumeIsolated()
                 .map { result -> ProducerElement in
-                    context.eventLoop.preconditionInEventLoop()
                     // We have to fire through the original data now. Since our channelReadTransformation
                     // is the channel initializer. Once that's done we need to fire the channel as a read
                     // so that it hits channelRead0 in the base socket channel.
                     context.fireChannelRead(data)
                     return result
                 }
+                .assumeIsolated()
```

Comment:
FWIW in this case we could just ditch the `map` and switch on the result in `whenComplete`